compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2716,76 +2717,203 @@ public void visit(BLangInvocation iExpr) {
     }
 
     public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
-        String typeName = "";
-        String pkgAlias = "";
-        if (errorConstructorExpr.errorTypeRef == null) {
-            typeName = "error";
-        } else {
-            typeName = errorConstructorExpr.errorTypeRef.typeName.value;
-            pkgAlias = errorConstructorExpr.errorTypeRef.pkgAlias.value;
+        BLangUserDefinedType userProvidedTypeRef = errorConstructorExpr.errorTypeRef;
+        BLangUserDefinedType errorTypeRef = userProvidedTypeRef;
+        if (errorTypeRef != null) {
+            symResolver.resolveTypeNode(errorTypeRef, env, DiagnosticErrorCode.UNDEFINED_ERROR_TYPE_DESCRIPTOR);
         }
+        validateErrorConstructorPositionalArgs(errorConstructorExpr);
 
-        BSymbol symbol = symResolver.lookupMainSpaceSymbolInPackage(errorConstructorExpr.pos, env,
-                names.fromString(pkgAlias), names.fromString(typeName));
-        if (symbol == symTable.notFoundSymbol || symbol.tag != SymTag.ERROR) {
-            dlog.error(errorConstructorExpr.pos, DiagnosticErrorCode.UNDEFINED_ERROR_TYPE_DESCRIPTOR, typeName);
-            resultType = symTable.semanticError;
-            return;
-        }
-        BErrorType errorConstructorType = (BErrorType) symbol.type;
-        if (this.expType == symTable.noType) {
-            this.expType = errorConstructorType;
+        List<BType> expandedCandidates = getTypeCandidatesForErrorConstructor(errorConstructorExpr);
+
+        List<BType> errorDetailTypes = new ArrayList<>();
+        for (BType expandedCandidate : expandedCandidates) {
+            BType detailType = ((BErrorType) expandedCandidate).detailType;
+            if (types.isAssignable(detailType, symTable.detailType)) {
```

Comment:
I was trying to avoid non error detail types here, as you pointed out in https://github.com/ballerina-platform/ballerina-lang/pull/29059/files#r607740063, we can't skip this here to avoid wrong index being pulled from the `expandedCandidates`